Supporting sign in fcode using f77

bjodah opened the pull request:
Fixes #12267, supersedes #12351 

- Introduce a new module: `.codegen.ffunctions`
- Intercept `Assignment` in fcode.py and check for sign

Pinging @Franjov for review and feedback.
Pinging @SatyaPrakashDwibedi since this PR is essentially feedback on how to extend the code-printers (in regards to #12351)

bjodah committed:
Supporting sign in fcode using f77

- Introduce new module .codegen.ffunctions
- Intercept Assignment in fcode and check for sign

SatyaPrakashDwibedi commented on sympy/printing/fcode.py:
```diff
@@ -178,6 +172,32 @@ def _print_Piecewise(self, expr):
+        return super(FCodePrinter, self)._print_Assignment(expr)
+
+    def _print_sign(self, expr):
+        if self._settings["standard"] < 95:
```
@bjodah 
I think you should add for version<95 as suggested by @Franjov by adding
```
if expr.has(Assignment):
#code here
```
as I did because now there won't be any tension about assign because as you have overrided now; and it's a good idea.

SatyaPrakashDwibedi reviewed (approved):
@bjodah  The Idea is awesome have a look at #12356 .Here I think I should edit `codeprinter` i.e the base class because it needs to be done for all.

Franjov commented on sympy/printing/fcode.py:
```diff
@@ -178,6 +172,32 @@ def _print_Piecewise(self, expr):
                                       "inline operators is not supported in "
                                       "standards earlier than Fortran95.")
 
+    def _print_Assignment(self, expr):
```
It would be clearer if the the content of `_print_Assignment` was into `_print_sign`.
What prevents you to use only one method?

SatyaPrakashDwibedi replied:
Because here @bjodah overrided the `_print_Assignment` in `codeprinter` i.e the base class.

Franjov replied:
Ok, I believe I found the answer in @SatyaPrakashDwibedi's comment in #12351.

Franjov replied:
Although if you replace `sp.sign` by some `sp.Piecewise` expression somewhere upwind, you might not need `_print_sign` anymore since `_print_Piecewise` is able to return the right `merge` construction for modern Fortran.

bjodah replied:
Well, since upwind in this case means Assignment, e.g.
`Assignment(Symbol('x'), sign(Symbol('x')))`
(the printer walks the tree and sees assignment before sing, and then the printing method of `Assignment` has a special case for `Piecewise` and now also `sign`). Therefore we will still need a `_print_sign` method for the cases when it is not part of an assignment. But I can refactor so that `_print_Assignment` and `_print_sign` share some private function with the logic for printing `sign` using `Piecewise`

asmeurer replied:
This is messy logic. What if the rhs contains sign as a subexpression? 

I take it Fortran doesn't allow assignments in expressions like C. 

We really need a good distinction (in all the code printers) between statements and expressions, and things that can only be printed as expressions (an expression being any piece of code that can be inserted arbitrarily in a larger expression, with the right parentheses).

bjodah replied:
I agree, but the code printers are already doing these kinds of hacks, and it's a bit of a slippery slope. I actually think it might be better to simply not support `sign` for Fortran standards < 95 since these fixes clutters the printers. Or maybe we should have a different printer class for writing statements and blocks thereof.

bjodah committed:
Refactor _print_sign logic in Fortran code printer

bjodah commented:
Ok, so I refactored. @Franjov in my last commit I changed the expected string in the test for printing `sign`.
I'm a bit unsure about implicit conversion rules in fortran, I don't think the user will loose any precision the way it's written now (`0` and `0d0` are both exactly zero), what do you say?

bjodah committed:
Replace sign with Piecewise early in Fortran printer

bjodah commented:
So I changed to approach once more, now I'm simply intercepting instances of `sign` in the `doprint` method of `FCodePrinter`, all complexities regarding `Assignment` are now handled by the already existing logic for handling `Piecewise`.

asmeurer commented:
Handling things in doprint seems even worse than handling them in _print_Assignment. 

We need some better way to abstract the notion of "Assignment(a, b) should be printed in a custom way, different from 'a = b', for certain b". I'm not really sure how to do it. But ideally it should allow extension to custom types, without subclasses having to copy-paste all of `_print_Assignment` to add some lines. 

Maybe the classes that split on assignment should have special `_print_assignment_Piecewise` method on the printer that tells it how to print it as an assignment. The question is how to handle this for subexpressions (like `Assignment(x, 1 + Piecewise(...))`).

bjodah commented:
I agree completely. We can simply defer this PR for now until that (major) restructuring of the code printers is in place. Should I close the PR or leave it open? (there is a reference to it in the issue)

asmeurer commented:
Since sign already works with the newer standards without any hacks I would prefer to defer. 

Although the fcode stuff seems useful. Do we need the printing to implement it?

bjodah commented:
I could open a separate PR which only brings `ffunctions.py` along (and refactors `_print_sign` in master to use those new classes).

Franjov commented:
That sounds reasonable. The Fortran 77 case is more complex than I thought, for a quite small gain.

@bjodah About precision of constants.
Personally, I would prefer to compare the real `x` to `0d0` rather than `0`.
Even if the compiler do not make any difference (to be checked), it might be more readable for the user to avoid any ambiguity about the type of the variable.
